parsers/fasta/fasta.go

```diff
@@ -0,0 +1,165 @@
+package fasta
+
+import (
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+/******************************************************************************
+Apr 25, 2021
+
+FASTA Parser begins here
+
+Many thanks to Jordan Campbell (https://github.com/0x106) for building the first
+FASTA parser for Poly. This work expands on that work by allowing for concurrent
+FASTA parsing and giving Poly a specific FASTA parser subpackage, as well as few
+bug fixes.
+
+FASTA is a very simple file format for working with DNA, RNA, or protein sequences.
+It was first released in 1985 and is still widely used in bioinformatics.
+
+https://en.wikipedia.org/wiki/FASTA_format
+
+One interesting use of the concurrent FASTA parser is working with the Uniprot
+FASTA dump files, which are far too large to fit into RAM. This parser is able
+to easily handle those files by doing computation actively while the data dump
+is getting parsed.
+
+https://www.uniprot.org/downloads
+
+I have removed the FASTA Parsers from the io.go file and moved them into this
+subpackage.
+
+Hack the Planet,
+
+Keoni
+
+******************************************************************************/
+
+// Fasta is a struct representing a single Fasta file element with a Name and its corresponding Sequence.
+type Fasta struct {
+	Name     string `json:"name"`
+	Sequence string `json:"sequence"`
+}
+
+// ParseFASTA parses a given Fasta file into an array of Fasta structs. Internally, it uses ParseFastaConcurrent.
+func ParseFASTA(r io.Reader) []Fasta {
+	fastas := make(chan Fasta, 1000) // A buffer is used so that the functions runs as it is appending to outputFastas
+	go ParseFASTAConcurrent(r, fastas)
+
+	var outputFastas []Fasta
+	for fasta := range fastas {
+		outputFastas = append(outputFastas, fasta)
+	}
+	return outputFastas
+}
+
+// ParseFASTAConcurrent concurrently parses a given Fasta file in an io.Reader into a channel of Fasta structs.
+func ParseFASTAConcurrent(r io.Reader, sequences chan<- Fasta) {
```

Comment:
The last parser was technically a multi-fasta parser that would add all Fastas into one struct. It would also annotate each segment based on its comment. Does this do that too?

Reply:
It looks like it doesn't interface with the Sequence struct at all which makes a lot of methods moot. Can this be fixed?

Reply:
> The last parser was technically a multi-fasta parser that would add all Fastas into one struct. It would also annotate each segment based on its comment. Does this do that too?

No, and it can't. That method rapidly inflated the byte size of the sequence struct, and we can't parse a 400Gb FASTA file because the memory requirement in the design of annotating each segment into a single sequence, meaning we'd have to break it up.

> It looks like it doesn't interface with the Sequence struct at all which makes a lot of methods moot. Can this be fixed?

No. For simplicity sake, FASTA files ARE different than Genbank files in non-trivial ways (for example, one has type information for seqhash, one does not). They should be able to share an interface, but I don't think they have or should have equivalent methods.